### PR TITLE
Decimal improvements

### PR DESCRIPTION
**Pull request by LukasGasior1**
Fix Decimal toString and toBigIntegerSubunits for negative numbers;
Add more tests;
Rename some methods to make it clear they only work for non-negative numbers

### CODE REVIEW

**Review by malexandru-rdx — APPROVED** [review]

File: common/src/test/java/com/radixdlt/utils/UInt192Test.java [hunk 1]
```diff
@@ -484,6 +484,48 @@ public void illegalArgumentExceptionOnNegativeExponent() {
...
+            .divide(UInt192.FOUR),
+        UInt192.from("784637716923335095479473677900958302012794430558004314112"));
+  }
+
+  /** Test some concrete large values add/sub */
+  @Test
+  public void test_large_number_add_sub() {
+    final var u1 = UInt192.from("3766261041232008458301473653924599849661413266678420707737");
+    // An overflow
+    assertEquals(
```
Comment (dhedey) [inline common/src/test/java/com/radixdlt/utils/UInt192Test.java:517]: Why does an overflow not throw an exception?

File: common/src/test/java/com/radixdlt/utils/UInt192Test.java [hunk 2]
```diff
@@ -484,6 +484,48 @@ public void illegalArgumentExceptionOnNegativeExponent() {
...
+        u1.add(u1), UInt192.from("1255420347077336152767157884641533283220471088892806902578"));
+
+    // A transient overflow
+    assertEquals(UInt192.MAX_VALUE.add(UInt192.ONE).subtract(UInt192.ONE), UInt192.MAX_VALUE);
```
Comment (dhedey) [inline common/src/test/java/com/radixdlt/utils/UInt192Test.java:521]: Ditto here?

File: core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java [hunk 3]
```diff
@@ -75,11 +75,31 @@
 import java.util.Objects;
 import org.bouncycastle.util.Arrays;
 
-/** Decimal represents a 192 bit representation of a fixed-scale decimal number. */
+/**
+ * Decimal represents a 192 bit representation of a fixed-scale decimal number. Note that the Java
+ * implementation is fairly basic and mostly acts as a container for node<->engine interop.
+ * Internally it uses an *unsigned* 192-bit integer representation, and only exposes a handful of
```
Comment (dhedey) [inline core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java:81]: But we're interpreting it as signed right? Can we make that clear?

File: core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java [hunk 4]
```diff
@@ -75,11 +75,31 @@
...
+ */
 @SecurityCritical(SecurityKind.NUMERIC)
 public class Decimal implements Comparable<Decimal> {
 
-  public static final Decimal ZERO = Decimal.of(0L);
+  public static final Decimal MIN_VALUE =
+      Decimal.fromUnsignedFixedPointRepresentation(
+          UInt192.from(
+              new BigInteger("-3138550867693340381917894711603833208051177722232017256448")
+                  .toByteArray()));
```
Comment (dhedey) [inline core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java:91]: I see we're relying on the fact that `.toByteArray()`  for `UInt192` and `BigInteger` is identical. But we have the DecimalTests which seem to test this well, so I guess it's fine to rely on this.

**Review by dhedey — APPROVED** [review]

File: core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java (hunk 4 above)
Comment (LukasGasior1) [reply core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java:91]: It doesn't always work for negative numbers (e.g. shorter than 24 bytes), but it does for MIN_VALUE. I'll add a comment to make that clear.

File: common/src/test/java/com/radixdlt/utils/UInt192Test.java (hunk 1 above)
Comment (LukasGasior1) [reply common/src/test/java/com/radixdlt/utils/UInt192Test.java:517]: Well, it doesn't by design. I don't think it's an issue on the java side?

File: core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java [hunk 5]
```diff
@@ -88,82 +109,90 @@ public static void registerCodec(CodecMap codecMap) {
...
+    return fromNonNegativeBigIntegerSubunits(
+        BigInteger.valueOf(Math.abs(numerator))
+            .multiply(BigInteger.TEN.pow(SCALE))
+            .divide(BigInteger.valueOf(Math.abs(denominator))));
   }
 
-  public Decimal subtract(Decimal other) {
-    var newUnderlying = this.underlyingValue.subtract(other.underlyingValue);
-    return new Decimal(newUnderlying);
+  public Decimal add(Decimal other) {
```
Comment (dhedey) [inline core-rust-bridge/src/main/java/com/radixdlt/rev2/Decimal.java]: This should be renamed to wrappingAdd etc

**Review by dhedey — APPROVED** [review]
Okay I've done the mark-ups as we need to get this out today, and will merge when the tests pass. Please do a post-review tomorrow of my commit when you're able.

**Pull request merged**